web-common/src/main/java/edu/unc/lib/boxc/web/common/services/AccessCopiesService.java

```diff
@@ -103,74 +109,99 @@ public boolean hasViewableFiles(ContentObjectRecord briefObj, AccessGroupSet pri
-        ContentObjectRecord contentObj = getContentObject(contentObjectRecord, principals);
-        if (contentObj != null) {
-            return DatastreamUtil.getOriginalFileUrl(contentObj);
+    public String getThumbnailId(ContentObjectRecord contentObjectRecord, AccessGroupSet principals,
+                                 boolean checkChildren) {
+        // Find thumbnail datastream recorded directly on the object, if present
+        var thumbId = DatastreamUtil.getThumbnailOwnerId(contentObjectRecord);
+        if (thumbId != null) {
+            log.debug("Found thumbnail object directly assigned to object {}", thumbId);
+            return thumbId;
+        }
+        // Don't need to check any further if object isn't a work or doesn't contain files with thumbnails
+        if (!ResourceType.Work.name().equals(contentObjectRecord.getResourceType())
+                || contentObjectRecord.getContentType() == null
+                || !contentObjectRecord.getContentType().contains(IMAGE_CONTENT_TYPE)) {
+            log.debug("Record {} is not applicable for a thumbnail", contentObjectRecord.getId());
+            return null;
+        }
+        if (!checkChildren) {
+            log.debug("Not checking children for work {}, so using self as thumbnail id", contentObjectRecord.getId());
+            return contentObjectRecord.getId();
+        }
+
+        SolrQuery query = buildFirstChildQuery(contentObjectRecord, principals);
+        // Limit query to just children which have a thumbnail datastream
+        query.addFilterQuery(solrSettings.getFieldName(SearchFieldKey.DATASTREAM.name()) + ":"
+                + DatastreamType.THUMBNAIL_LARGE.getId() + "|*");
+
+        try {
+            QueryResponse resp = executeQuery(query);
+            if (resp.getResults().getNumFound() > 0) {
+                var results = resp.getBeans(ContentObjectSolrRecord.class);
+                var id = results.get(0).getId();
+                log.debug("Found thumbnail object {} for work {}", id, contentObjectRecord.getId());
+                return results.get(0).getId();
+            } else {
+                log.debug("No thumbnail objects for work {}", contentObjectRecord.getId());
+                return null;
+            }
+        } catch (SolrServerException e) {
+            throw new SolrRuntimeException("Error listing viewable files: " + query, e);
         }
+    }
+
+    public void populateThumbnailId(ContentObjectRecord record, AccessGroupSet principals,
+                                    boolean checkChildren) {
+        if (record == null) {
+            return;
+        }
+        record.setThumbnailId(getThumbnailId(record, principals, checkChildren));
+    }
 
-        return "";
+    public void populateThumbnailIds(List<ContentObjectRecord> records, AccessGroupSet principals,
+                                     boolean checkChildren) {
+        for (var record: records) {
```

Comment:
Should there be a space after record?

Reply:
I keep forgetting whether we do spaces there or not, i'll add one